Persistent Grades Flag multiple entries per course

efischer19 opened the pull request:
With this change, it is now possible to enable and disable the feature for a
particular course as many times as desired. Fix confirmed locally.

TNL-5698

efischer19 commented on lms/djangoapps/grades/migrations/0005_multiple_course_flags.py:
```diff
@@ -0,0 +1,20 @@
+            name='course_id',
+            field=xmodule_django.models.CourseKeyField(max_length=255, db_index=True),
+        ),
+    ]
```
@jibsheet Are there any devops concerns with dropping an index in a migration? This table is empty on prod, and has 4 entries on stage.

```
edxapp@precise64:~/edx-platform$ ./manage.py lms --settings=devstack sqlmigrate grades 0005
BEGIN;
ALTER TABLE `grades_coursepersistentgradesflag` DROP INDEX `grades_coursepersistentgradesfla_course_id_539df2575c05593b_uniq`;

COMMIT;
```

nasthagiri replied:
It there a migration code change where it removes the uniqueness constraint?

efischer19 replied:
Yes, that's what this migration does. See the output of `sqlmigrate` I posted above.

jibsheet replied:
@efischer19 I was out, but to confirm (on this merged PR) dropping this index on a table this small is fine.

efischer19 commented on lms/djangoapps/grades/config/tests/test_models.py:
```diff
@@ -47,3 +47,22 @@ def test_persistent_grades_feature_flags(self, global_flag, enabled_for_all_cour
+            course_id=self.course_id_1,
+            enabled_for_course=True
+        ):
+            # Prior to TNL-5698, creating a second object would fail due to db constraints
```
I've confirmed this test to fail on the old version of the code.

```
ERROR: test_enable_disable_course_flag (lms.djangoapps.grades.config.tests.test_models.PersistentGradesFeatureFlagTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mock.py", line 1618, in _inner
    return f(*args, **kw)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/config/tests/test_models.py", line 66, in test_enable_disable_course_flag
    enabled_for_course=False
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/config/tests/utils.py", line 23, in persistent_grades_feature_flags
    CoursePersistentGradesFlag.objects.create(course_id=course_id, enabled=enabled_for_course)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/query.py", line 348, in create
    obj.save(force_insert=True, using=self.db)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/config_models/models.py", line 110, in save
    super(ConfigurationModel, self).save(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 734, in save
    force_update=force_update, update_fields=update_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 762, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 846, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py", line 885, in _do_insert
    using=using, raw=raw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/query.py", line 920, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 974, in execute_sql
    cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py", line 318, in execute
    return Database.Cursor.execute(self, query, params)
IntegrityError: column course_id is not unique
```

nasthagiri replied:
Add a `self.assertTrue` to verify that the flag is enabled for the course at this point.

efischer19 replied:
Thanks for requesting this - revealed an error in the test not testing what I thought (`global_flag` should have been True in the context manager), which once fixed revealed a problem in `feature_enabled()` (`get()` fails w/ more than 1 result)

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run a11y

efischer19 commented:
@jcdyer More potentially flaky tests:

a11y:
- https://build.testeng.edx.org/job/edx-platform-accessibility-pr/25493/testReport/junit/common.test.acceptance.tests.studio.test_studio_textbooks/TextbooksTest/test_pdf_viewer_a11y/

lettuce:
- https://build.testeng.edx.org/job/edx-platform-lettuce-pr/26181/testReport/junit/LMS/Answer%20problems%20_%20I%20can%20view%20and%20hide%20the%20answer%20if%20the%20problem%20has%20it_/Then_the_Show_Hide_button_label_is__HIDE_ANSWER_/
- https://build.testeng.edx.org/job/edx-platform-lettuce-pr/26181/testReport/junit/LMS/Answer%20problems%20_%20I%20can%20reset%20the%20correctness%20of%20a%20problem%20after%20changing%20my%20answer/__correct___formula___incorrect__/

efischer19 commented:
@nasthagiri @jcdyer @sanfordstudent This is ready for review, I'll mention it in scrum.

efischer19 commented:
Bulk email does have the same issue; I'll file a separate PR for that one.

sanfordstudent commented:
I know I already approved this but good updates. I missed that `global_flag` was set to `False` the first time around. Should we have a test that we can disable the flag for all courses (i.e. `enabled_for_all_courses` can be set to false after being set to true) as well?

efischer19 commented:
@sanfordstudent we have one now!

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
@jcdyer more possibly flaky tests:
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/27350/testReport/junit/common.test.acceptance.tests.lms.test_lms_edxnotes/EdxNotesPageTest/test_easy_access_from_notes_page/
- https://build.testeng.edx.org/job/edx-platform-lettuce-pr/26235/testReport/junit/LMS/Answer%20problems%20_%20The%20reset%20button%20doesn't%20show%20up/__never___correct___formula__/

nasthagiri commented:
👍 once squashed, etc.

efischer19 commented:
Squashed. @sanfordstudent, since your last comment I added the [test_enable_disable_globally test](https://github.com/edx/edx-platform/blob/78075d8bf6c8991a934179733ab5f297dc6fb868/lms/djangoapps/grades/config/tests/test_models.py#L71), can you thumbs that change before I merge?

sanfordstudent commented:
👍 Thanks!

efischer19 commented:
jenkins run lettuce

lamagnifica commented:
@efischer19 will this have any effect on the data packages?

efischer19 commented:
@lamagnifica no, it won't. The only functional change here was to remove an index that was preventing the feature from functioning as originally intended.
